File: apps/api/app/models/action_item.py
import uuid
from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base


class ActionItem(Base):
    __tablename__ = "action_items"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    meeting_id: Mapped[str] = mapped_column(ForeignKey("meetings.id"), nullable=False, index=True)
    task: Mapped[str] = mapped_column(Text, nullable=False)
    owner_name: Mapped[str | None] = mapped_column(String)
    owner_email: Mapped[str | None] = mapped_column(String, index=True)
    deadline: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    priority: Mapped[str] = mapped_column(String, default="medium", nullable=False)
    status: Mapped[str] = mapped_column(String, default="open", nullable=False, index=True)
    verbatim_quote: Mapped[str | None] = mapped_column(Text)
    resolved_in_meeting_id: Mapped[str | None] = mapped_column(ForeignKey("meetings.id"))
    # Stored as hex string of float array for pgvector compatibility via raw SQL
    embedding_id: Mapped[str | None] = mapped_column(String)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )

    meeting: Mapped["Meeting"] = relationship(back_populates="action_items", foreign_keys=[meeting_id])  # noqa: F821

File: packages/shared/src/types/action-item.ts
export type ActionItemStatus =
  | "open"
  | "in_progress"
  | "done"
  | "overdue"
  | "deferred";

export type ActionItemPriority = "low" | "medium" | "high" | "critical";

export interface ActionItem {
  id: string;
  meetingId: string;
  task: string;
  ownerName?: string;
  ownerEmail?: string;
  deadline?: string;
  priority: ActionItemPriority;
  status: ActionItemStatus;
  verbatimQuote?: string;
  resolvedInMeetingId?: string;
  embeddingId?: string;
  createdAt: string;
  updatedAt: string;
}
